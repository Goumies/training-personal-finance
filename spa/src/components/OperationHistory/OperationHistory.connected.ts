import { connect, State, Updater } from "../../store";
import { OperationHistoryComponent, StateProps, UpdateProps, OwnProps } from "./OperationHistory";
import { Operation } from "../../interfaces/Operation";

function computeStateProps({ operations }: State): StateProps {
  return { operations };
}

function computeUpdateProps(updateState: (updater: Updater) => void): UpdateProps {
  function onInitOperations(operations: Operation[]) {
    const updater: Updater = (state: State) => {
      return {
        ...state, 
        operations
      }
    }
    updateState(updater);
  };
  return { onInitOperations };
}

export const OperationHistory = connect<OwnProps, StateProps, UpdateProps>(
  OperationHistoryComponent,
  { computeStateProps, computeUpdateProps }
);
