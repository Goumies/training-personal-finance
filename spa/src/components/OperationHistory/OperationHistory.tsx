import React from "react";
import { Operation as IOperation } from "../../interfaces/Operation";
import { Operation } from "../Operation";
import nodeFetch from "node-fetch";
import { render } from "enzyme";

export interface OwnProps {}

export interface StateProps {
  operations: IOperation[];
}

export interface UpdateProps {
  onInitOperations(operations: IOperation[]): void;
}

export type Props = OwnProps & StateProps & UpdateProps;

export class OperationHistoryComponent extends React.Component<Props, {}> {
  constructor(props: Props) {
    super(props);
  }

  componentDidMount() {
    nodeFetch('http://localhost:5000/operations')
    .then(res => res.json())
    .then(json => {
      const seriallizedOperations: { id: string, date: string, amount: number }[] = json;
      const deserializedOperations = seriallizedOperations
        .map(operation => ({
          ...operation,
          date: new Date(operation.date)
        }));
      this.props.onInitOperations(deserializedOperations);
    });
  }

  render() {
    return (
      <div>
        {this.props.operations.map((operation: IOperation) => (
          <Operation key={operation.id} id={operation.id} />
        ))}
      </div>
    );
  }
    
};
