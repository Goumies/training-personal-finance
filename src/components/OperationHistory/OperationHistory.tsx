import React from "react";
import { Operation as IOperation } from "../../interfaces/Operation";
import { Operation } from "../Operation";

interface Props {
  operations: IOperation[];
}

export const OperationHistory: React.StatelessComponent<Props> = ({ operations }: Props) => {
  return (
    <div>
      {operations.map((operation: IOperation) => (
        <Operation key={operation.id} operation={operation} />
      ))}
    </div>
  );
};
