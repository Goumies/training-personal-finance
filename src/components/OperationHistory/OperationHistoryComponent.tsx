import React from "react";
import { Operation } from "../../interfaces/Operation";
import { OperationComponent } from "../Operation";

interface Props {
  operations: Operation[];
}

export const OperationHistoryComponent: React.StatelessComponent<Props> = ({ operations }: Props) => {
  return (
    <div>
      {operations.map((operation: Operation) => (
        <OperationComponent key={operation.id} operation={operation} />
      ))}
    </div>
  );
};
