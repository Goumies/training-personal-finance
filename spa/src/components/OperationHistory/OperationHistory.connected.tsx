import React from "react";
import { StoreContext, Store } from "../../store";
import { OperationHistoryComponent, OwnProps } from "./OperationHistory";

export const OperationHistory: React.StatelessComponent<OwnProps> = () => (
  <StoreContext.Consumer>
    {(store: Store | undefined) => {
      const operations = store === undefined ? [] : store.getState().operations;
      return <OperationHistoryComponent operations={operations} />;
    }}
  </StoreContext.Consumer>
);
