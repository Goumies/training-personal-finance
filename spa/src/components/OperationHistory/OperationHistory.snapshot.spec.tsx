import React from "react";
import renderer from "react-test-renderer";
import { OperationHistory, Props } from "./OperationHistory";

it("renders correctly", () => {
  const props: Props = {
    operations: [
      {
        id: "0",
        date: new Date(2019, 4, 24),
        amount: 10
      },
      {
        id: "1",
        date: new Date(2019, 4, 25),
        amount: -5
      }
    ]
  };
  const tree = renderer.create(<OperationHistory {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});
