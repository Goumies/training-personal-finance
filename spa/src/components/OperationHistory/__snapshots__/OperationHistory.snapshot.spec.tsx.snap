// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="operation"
    data-e2e="account-operation"
  >
    <div
      className="labeled-debit-or-credit"
    >
      <div
        className="label-container"
      >
        <span
          className="date"
        >
          2019-5-24
        </span>
      </div>
      <div
        className="amount-container"
      >
        <div
          className="debit-or-credit"
        >
          <span
            className="credit"
          >
            <span
              className="value"
            >
              +10
            </span>
          </span>
          <span
            className="debit"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="operation"
    data-e2e="account-operation"
  >
    <div
      className="labeled-debit-or-credit"
    >
      <div
        className="label-container"
      >
        <span
          className="date"
        >
          2019-5-25
        </span>
      </div>
      <div
        className="amount-container"
      >
        <div
          className="debit-or-credit"
        >
          <span
            className="credit"
          />
          <span
            className="debit"
          >
            <span
              className="value"
            >
              -5
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
